### PR TITLE
Add design details for new filter in handler chain

### PR DESCRIPTION
**Pull request by richabanker**
- One-line PR description: amends to UVIP KEP

### CODE REVIEW

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 1]
```diff
@@ -275,20 +275,48 @@ with a metric.
 
 There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade -- that would cause other problems, too.
 
-TODO: security / cert stuff.
+We should ensure a definite way for the request flow to complete rather than continuing on indefinitely by proxying the request over and over again.
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]:
```suggestion
We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve).
```

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 2]
```diff
@@ -275,20 +275,48 @@ with a metric.
...
+
+We need to watch out for inadvertantly putting apiserver IP/endpoint information and the trust bundle control (used to authenticate server while proxying) in user hands (by exposing this information in REST APIs)
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: The risk here is that users could modify it and then do a MitM attack?

"in user hands" is a metaphor that isn't very clear.

I would probably rephrase this as "To prevent {failure mode X} we will not {do thing that could lead to that}".

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 3]
```diff
@@ -275,20 +275,48 @@ with a metric.
...
+    1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly.
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+    1. It is a non resource request
+    2. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+    3. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver.
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 503 with error `No relevant API server found for the requested GVR <group_version_resource>`
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: hm, if no apiserver serves it, wouldn't 404 be correct, as we know for sure that it is absent?

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 4]
```diff
@@ -275,20 +275,48 @@ with a metric.
...
-TODO: security / cert stuff.
+We should ensure a definite way for the request flow to complete rather than continuing on indefinitely by proxying the request over and over again.
+
+We need to watch out for inadvertantly putting apiserver IP/endpoint information and the trust bundle control (used to authenticate server while proxying) in user hands (by exposing this information in REST APIs)
 
 ## Design Details
 
-TODO: explanation of how the handler will determine a request is for a resource
-that should be proxied.
+### Aggregation Layer
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md:284]: At some point in this text, can you add a description of what happens before the map is populated?

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 5]
```diff
@@ -275,20 +275,48 @@ with a metric.
...
+    2. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+    3. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver.
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 503 with error `No relevant API server found for the requested GVR <group_version_resource>`
+
+[handler chain]:https://github.com/kubernetes/kubernetes/blob/fc8f5a64106c30c50ee2bbcd1d35e6cd05f63b00/staging/src/k8s.io/apiserver/pkg/server/config.go#L639
+
+In order to ensure that we do not end up rerouting the request indefinitely from one apiserver to another, we will ensure that we short circuit an already proxied request to be served locally, by adding a header to the proxied request. Then, when the destination apiserver finds this header in the request, we should not proxy the request again.
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: You could add this as an item under step 2 above?

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 6]
```diff
@@ -275,20 +275,48 @@ with a metric.
...
+* TODO: We need to find a place to store and retrieve the destination apiserver's host and port information given the server's ID.
+We do not want to store this information in
 
-TODO: explanation of how the security handshake between apiservers works.
-* What we need to fix: random processes / external users / etc should not be
-  able to proxy requests, so the receiving apiserver needs to be able to verify
-  the source apiserver.
-* generate self-signed cert on startup, put pubkey in apiserver identity lease
-  object?
+    * StorageVersion : because we do not want to expose the network identity of the apiservers in this API that can be listed in places where it could be unsafe to do so
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: The problem is it would be highly redundant, it's not about safety

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 3 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: But if the storageversion for the gvr was found, could the missing apiserver ID be because of some delay in updating the records? Is it correct to serve 404 still?

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 2 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Rephrased as suggested.

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 4 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md:284]: Added text about waiting for informer cache to sync + wait for StorageVersionManager to return true. Anything else that I am missing?

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 5 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Done.

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 6 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Corrected, thanks!

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 1 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Something happened here, I applied your commit but it got lost (I think I forgot to pull before pushing), I have made this change manually now though.

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 3 above)
Comment (lavalamp) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: I think a storage version with no entries should be treated the same as not having a storage version at all. (404)

But a storage version which has only entries but those entries refer to apiservers that are not healthy or unresponsive should get 503s, since maybe they will come back. (In practice we don't need to handle this specially, we're not health checking them, we just attempt the proxy and if the proxy fails we return a 503)

Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Got it. For the case when there is no storage version found, we pass the request to the next handler in local apiserver, since it could be a request for an aggregated API or for a CR

But in the case of not finding any apiserver, should we still pass on the request as it is to the next handler or fail fast with 404?

Comment (lavalamp) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: I think it will get a 404 either way, so it doesn't really matter.

(I assume we already detected CRs and AAs and let them proceed down the handler chain)

Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Okay, updated text to serve 404 in this case. Thanks for correcting my understanding!

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 7]
```diff
@@ -275,20 +275,57 @@ with a metric.
 
 There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade -- that would cause other problems, too.
 
-TODO: security / cert stuff.
+We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve).
+
+To prevent Man in the Middle attack we will not expose information about apiserver IP/endpoint and the trust bundle control (used to authenticate server while proxying) in REST APIs.
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: The problem isn't "exposing" it there -- the trust bundle is already distributed to all clients -- the problem is permitting someone with api access to inject a malicious one.

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 8]
```diff
@@ -275,20 +275,57 @@ with a metric.
...
-TODO: explanation of how the handler will determine a request is for a resource
-that should be proxied.
+### Aggregation Layer
+
+1. A new filter will be added to the [handler chain] of the aggregation layer. This filter will maintain an internal map with the key being the group-version-resource and the value being a list of server IDs of apiservers that are capable of serving that group-version-resource
+  1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+  1. It is a non resource request
+  2. The StorageVersion informer cache hasn't synced yet or if `StorageVersionManager.Completed()` has returned false
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: We'll need to modify something further down the chain to take special care to serve 503 instead of 404 in this case, if the resource actually isn't served locally.

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 9]
```diff
@@ -275,20 +275,57 @@ with a metric.
...
+  3. The request has a header that indicates that this request has been proxied once already
+  4. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+  5. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 404 with error `No relevant API server found for the requested GVR <group_version_resource>`
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: This case shouldn't happen, but I suggest the message should be "GVR is not served by anything in this cluster", i.e. it's a fact about the cluster, not an error condition about not finding an apiserver.

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 8 above)
Comment (lavalamp) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: We also need this property for the following point (it's been proxied once already but we don't actually serve it).

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md [hunk 10]
```diff
@@ -275,20 +275,57 @@ with a metric.
...
+
+5. If the proxy call fails for network issues or any reason, we serve 503 with error `Error while proxying request to destination apiserver`
+
+6. We will also add a poststarthook for the apiserver to perform the following checks to ensure that the server only starts serving requests after:
```
Comment (lavalamp) [inline keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: This is already addressed by item 2.2, actually not serving before this point would make self requests fail and then the server wouldn't be able to start at all. I would omit this point completely

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 7 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Modified, is it better now?

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 8 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Done

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 9 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Done

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 10 above)
Comment (richabanker) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: Removed

File: keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md (hunk 7 above)
Comment (lavalamp) [reply keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md]: yeah!

**Comment by lavalamp** [comment]
Thank you!
